provider/pkg/provider/provider.go

```diff
@@ -1077,6 +1102,30 @@ func (p *cfnProvider) Update(ctx context.Context, req *pulumirpc.UpdateRequest)
 	var outputs resource.PropertyMap
 	id := req.GetId()
 	resourceToken := string(urn.Type())
+
+	if req.GetPreview() {
+		if customResource, ok := p.customResources[resourceToken]; ok {
+			outputs = customResource.PreviewCustomResourceOutputs()
+		} else {
+			spec, hasSpec := p.resourceMap.Resources[resourceToken]
+			if !hasSpec {
+				return nil, errors.Errorf("Resource type %s not found", resourceToken)
+			}
+			resourceTypeName := urn.Type().Name()
+			outputs = previewResourceOutputs(newInputs, p.resourceMap.Types, spec.Outputs, spec.ReadOnly)
+			outputs = populateStableOutputs(newInputs, oldState, spec.ReadOnly, resourceTypeName)
```

Comment:
I still don't understand this code because when read naively it seems that populateStableOutputs discards the result of previewResourceOutputs and overwrites outputs. 

Should it be instead something like this?

```populateStableOutputs(newInputs, oldState, spec.ReadOnly, resourceTypeName, &outputs)```

I expect populateStableOutputs complement the work that previewResourceOutputs has done to keep populating more outputs (rewriting)?

Reply:
yes this was a bug, populateStableOutputs was supposed to take the result of previewResourceOutputs. I went ahead and refactored this so that it all happens in a single function.